File: Data Structures & Algorithms/binary-tree-diameter/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxVal = 0
    diameterOfBinaryTree(root){
        this.maxDepth_1(root)
        return this.maxVal;
    }

    maxDepth_1(root) {
        if(!root){
            return 0
        }
        let left = this.maxDepth_1(root.left)
        let right = this.maxDepth_1(root.right)
        if(left || right){
            this.maxVal = Math.max(this.maxVal, left + right)
        }
        return Math.max(left, right) + 1
    }
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        maxVal = [float("-inf")]
        def dfs(root):
            if(not(root)):
                return 0
            
            left = dfs(root.left)
            right = dfs(root.right)
            left = max(left, 0)
            right = max(right, 0)
            if(left + right + root.val > maxVal[0]):
                maxVal[0] = left + right + root.val
            return max(left, right) + root.val
        dfs(root)
        return maxVal[0]
